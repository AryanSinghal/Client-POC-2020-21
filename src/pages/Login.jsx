import React, { Component } from 'react';
import { Grid, InputAdornment, Typography, TextField, Card, CardContent, CardActions, Button } from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import EmailIcon from '@material-ui/icons/Email';
import VisibilityOffIcon from '@material-ui/icons/VisibilityOff';
import LockIcon from '@material-ui/icons/Lock';
import { callAPI } from '../libs/utils/api';

const style = {
  grid: { height: '90vh', paddingTop: "100px" },
  card: { width: '400px' },
  avatar: { backgroundColor: 'red' },
  textField: { width: "300px" },
  button: { width: "300px", align: "center" }
};

class Login extends Component {
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      password: '',
      passwordError: '',
      emailError: ''
    };
  }

  handleSubmit = () => {
    const { email, password } = this.state;
    console.log(email, password);
    const url = 'user';
    callAPI({ method: 'post', url, body: { email, password } })
      .then(({ data }) => {
        if (data) {
          console.log(data)
          localStorage.setItem('token', data);
          // this.props.history.push('');
        }
      })
      .catch((err) => {
        console.log(err.message);
      })
  }

  handlePasswordChange = (event) => {
    const { value } = event.target;
    (value.length >= 4)
      ? this.setState({ password: value, passwordError: '' })
      : this.setState({ password: value, passwordError: 'Password is required' })
  }

  handleEmailChange = (event) => {
    const { value } = event.target;
    const regex = /[a-z]([[-]*\w+[.]*){1,63}@[a-z]{2,63}[.][a-z]{2,63}$/gmi;
    if (value.length === 0) return this.setState({ email: value, emailError: 'Email is required' });
    (value.match(regex))
      ? this.setState({ email: value, emailError: '' })
      : this.setState({ email: value, emailError: 'Incorrect EmailId' });
  }

  render() {
    const { classes } = this.props;
    const { emailError, passwordError } = this.state;
    return (
      <div align="center" style={{ paddingTop: "100px" }}>
        <Card className={classes.card}>
          <Grid container direction="column" justify="center" alignItems="center" spacing={1}>
            <Grid item>
              <CardContent>
                <Grid container direction="column" justify="center" alignItems="center" spacing={3}>
                  <Grid item>
                    <Typography className='login' variant="h3">
                      Login
                    </Typography>
                  </Grid>
                  <Grid item>
                    <LockIcon />
                  </Grid>
                  <Grid item>
                    <TextField
                      className={classes.textField}
                      required
                      id="email"
                      label="Email"
                      variant="outlined"
                      fullWidth
                      helperText={emailError}
                      error={!!(emailError)}
                      InputProps={{
                        startAdornment: (
                          <InputAdornment position="start">
                            <EmailIcon />
                          </InputAdornment>
                        ),
                      }}
                      onChange={this.handleEmailChange}
                      onBlur={this.handleEmailChange}
                    />
                  </Grid>
                  <Grid item>
                    <TextField
                      className={classes.textField}
                      required
                      fullWidth
                      helperText={passwordError}
                      error={!!(passwordError)}
                      id="password"
                      label="Password"
                      type="password"
                      autoComplete="current-password"
                      InputProps={{
                        startAdornment: (
                          <InputAdornment position="start">
                            <VisibilityOffIcon />
                          </InputAdornment>
                        ),
                      }}
                      variant="outlined"
                      onChange={this.handlePasswordChange}
                      onBlur={this.handlePasswordChange}
                    />
                  </Grid>
                </Grid>
              </CardContent>
            </Grid>
            <Grid item>
              <CardActions>
                <Typography gutterBottom>
                  <Button
                    fullWidth
                    className={classes.button}
                    variant="contained"
                    color="primary"
                    size="medium"
                    align="center"
                    onClick={this.handleSubmit}
                  >
                    Sign In
                </Button>
                </Typography>
              </CardActions>
            </Grid>
          </Grid>
        </Card>
      </div>
    );
  }
}

export default withStyles(style)(Login);
export { Login };

