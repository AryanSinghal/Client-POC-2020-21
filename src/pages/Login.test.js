import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import { Login } from './Login';

Enzyme.configure({ adapter: new Adapter() });

describe('<Login />', () => {
  let wrapper;
  beforeAll(() => {
    const props = {
      classes: {
        grid: { height: '90vh', paddingTop: "100px" },
        card: { width: '400px' },
        avatar: { backgroundColor: 'red' },
        textField: { width: "300px" },
        button: { width: "300px", align: "center" }
      }
    };
    wrapper = shallow(<Login {...props} />);
  });

  it('should match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should not throw on correct Email', () => {
    const instance = wrapper.instance();
    instance.handleEmailChange({ target: { value: 'correctEmail@gmail.com' } });
    wrapper.update();
    expect(wrapper.state('emailError')).toBe('');
  });

  it('should show on incorrect Email', () => {
    const instance = wrapper.instance();
    instance.handleEmailChange({ target: { value: 'correctEmail@com' } });
    wrapper.update();
    expect(wrapper.state('emailError')).toBeTruthy();
  });

  it('should not throw on correct Password', () => {
    const instance = wrapper.instance();
    instance.handlePasswordChange({ target: { value: 'correctPassword' } });
    wrapper.update();
    expect(wrapper.state('passwordError')).toBe('');
  });

  it('should show error on incorrect Password', () => {
    const instance = wrapper.instance();
    instance.handlePasswordChange({ target: { value: 'c' } });
    wrapper.update();
    expect(wrapper.state('passwordError')).toBe('Password is required');
  });

  // it('should run handleSubmit', async () => {
  //   const instance = wrapper.instance();
  //   instance.setState({ email: 'correctEmail@gmail.com', password: 'correctPassword' })
  //   instance.handleSubmit();
  //   await new Promise(resolve => setTimeout(resolve, 0));
  //   wrapper.update();
  //   expect(localStorage.getItem('data')).toBeFalsy();
  // });

  it('', async () => {
    const instance = wrapper.instance();
    instance.setState({ email: 'correctEmail@gmail.com', password: 'correctPassword' });
    instance.handleSubmit = jest.fn().mockResolvedValue({});
    instance.handleSubmit();
    await new Promise(resolve => setTimeout(resolve, 0));
    const mutationCall = (data) => new Promise((resolve, reject) => ((data) ? resolve({ data })
      : reject()));
    wrapper.update();
    expect(localStorage.getItem('data')).toBeFalsy();
  });
});
