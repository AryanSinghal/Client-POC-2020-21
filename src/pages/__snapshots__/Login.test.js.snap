// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> renders a <Card /> component 1`] = `"<div align=\\"center\\" style=\\"padding-top:100px\\"><div class=\\"MuiPaper-root MuiCard-root width MuiPaper-elevation1 MuiPaper-rounded\\"><div class=\\"MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center\\"><div class=\\"MuiGrid-root MuiGrid-item\\"><div class=\\"MuiCardContent-root\\"><div class=\\"MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center\\"><div class=\\"MuiGrid-root MuiGrid-item\\"><h3 class=\\"MuiTypography-root login MuiTypography-h3\\">Login</h3></div><div class=\\"MuiGrid-root MuiGrid-item\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zm-6 9c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2zm3.1-9H8.9V6c0-1.71 1.39-3.1 3.1-3.1 1.71 0 3.1 1.39 3.1 3.1v2z\\"></path></svg></div><div class=\\"MuiGrid-root MuiGrid-item\\"><div class=\\"MuiFormControl-root MuiTextField-root width MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-required Mui-required\\" data-shrink=\\"true\\" for=\\"email\\" id=\\"email-label\\">Email<span class=\\"MuiFormLabel-asterisk MuiInputLabel-asterisk\\"> *</span></label><div class=\\"MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart MuiOutlinedInput-adornedStart\\"><div class=\\"MuiInputAdornment-root MuiInputAdornment-positionStart\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4l-8 5-8-5V6l8 5 8-5v2z\\"></path></svg></div><input type=\\"text\\" aria-invalid=\\"false\\" id=\\"email\\" required=\\"\\" class=\\"MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart\\"/><fieldset aria-hidden=\\"true\\" class=\\"PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline\\"><legend class=\\"PrivateNotchedOutline-legendLabelled-3 PrivateNotchedOutline-legendNotched-4\\"><span>Email *</span></legend></fieldset></div></div></div><div class=\\"MuiGrid-root MuiGrid-item\\"><div class=\\"MuiFormControl-root MuiTextField-root width MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-required Mui-required\\" data-shrink=\\"true\\" for=\\"password\\" id=\\"password-label\\">Password<span class=\\"MuiFormLabel-asterisk MuiInputLabel-asterisk\\"> *</span></label><div class=\\"MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart MuiOutlinedInput-adornedStart\\"><div class=\\"MuiInputAdornment-root MuiInputAdornment-positionStart\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M12 7c2.76 0 5 2.24 5 5 0 .65-.13 1.26-.36 1.83l2.92 2.92c1.51-1.26 2.7-2.89 3.43-4.75-1.73-4.39-6-7.5-11-7.5-1.4 0-2.74.25-3.98.7l2.16 2.16C10.74 7.13 11.35 7 12 7zM2 4.27l2.28 2.28.46.46C3.08 8.3 1.78 10.02 1 12c1.73 4.39 6 7.5 11 7.5 1.55 0 3.03-.3 4.38-.84l.42.42L19.73 22 21 20.73 3.27 3 2 4.27zM7.53 9.8l1.55 1.55c-.05.21-.08.43-.08.65 0 1.66 1.34 3 3 3 .22 0 .44-.03.65-.08l1.55 1.55c-.67.33-1.41.53-2.2.53-2.76 0-5-2.24-5-5 0-.79.2-1.53.53-2.2zm4.31-.78l3.15 3.15.02-.16c0-1.66-1.34-3-3-3l-.17.01z\\"></path></svg></div><input type=\\"password\\" aria-invalid=\\"false\\" autoComplete=\\"current-password\\" id=\\"password\\" required=\\"\\" class=\\"MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart\\"/><fieldset aria-hidden=\\"true\\" class=\\"PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline\\"><legend class=\\"PrivateNotchedOutline-legendLabelled-3 PrivateNotchedOutline-legendNotched-4\\"><span>Password *</span></legend></fieldset></div></div></div></div></div></div><div class=\\"MuiGrid-root MuiGrid-item\\"><div class=\\"MuiCardActions-root MuiCardActions-spacing\\"><p class=\\"MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom\\"><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained width align MuiButton-containedPrimary MuiButton-fullWidth\\" tabindex=\\"0\\" type=\\"button\\" align=\\"center\\"><span class=\\"MuiButton-label\\">Sign In</span></button></p></div></div></div></div></div>"`;

exports[`<Login /> should match snapshot 1`] = `"<div align=\\"center\\" style=\\"padding-top:100px\\"><div class=\\"MuiPaper-root MuiCard-root width MuiPaper-elevation1 MuiPaper-rounded\\"><div class=\\"MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center\\"><div class=\\"MuiGrid-root MuiGrid-item\\"><div class=\\"MuiCardContent-root\\"><div class=\\"MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center\\"><div class=\\"MuiGrid-root MuiGrid-item\\"><h3 class=\\"MuiTypography-root login MuiTypography-h3\\">Login</h3></div><div class=\\"MuiGrid-root MuiGrid-item\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zm-6 9c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2zm3.1-9H8.9V6c0-1.71 1.39-3.1 3.1-3.1 1.71 0 3.1 1.39 3.1 3.1v2z\\"></path></svg></div><div class=\\"MuiGrid-root MuiGrid-item\\"><div class=\\"MuiFormControl-root MuiTextField-root width MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-required Mui-required\\" data-shrink=\\"true\\" for=\\"email\\" id=\\"email-label\\">Email<span class=\\"MuiFormLabel-asterisk MuiInputLabel-asterisk\\"> *</span></label><div class=\\"MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart MuiOutlinedInput-adornedStart\\"><div class=\\"MuiInputAdornment-root MuiInputAdornment-positionStart\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4l-8 5-8-5V6l8 5 8-5v2z\\"></path></svg></div><input type=\\"text\\" aria-invalid=\\"false\\" id=\\"email\\" required=\\"\\" class=\\"MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart\\"/><fieldset aria-hidden=\\"true\\" class=\\"PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline\\"><legend class=\\"PrivateNotchedOutline-legendLabelled-3 PrivateNotchedOutline-legendNotched-4\\"><span>Email *</span></legend></fieldset></div></div></div><div class=\\"MuiGrid-root MuiGrid-item\\"><div class=\\"MuiFormControl-root MuiTextField-root width MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-required Mui-required\\" data-shrink=\\"true\\" for=\\"password\\" id=\\"password-label\\">Password<span class=\\"MuiFormLabel-asterisk MuiInputLabel-asterisk\\"> *</span></label><div class=\\"MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart MuiOutlinedInput-adornedStart\\"><div class=\\"MuiInputAdornment-root MuiInputAdornment-positionStart\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M12 7c2.76 0 5 2.24 5 5 0 .65-.13 1.26-.36 1.83l2.92 2.92c1.51-1.26 2.7-2.89 3.43-4.75-1.73-4.39-6-7.5-11-7.5-1.4 0-2.74.25-3.98.7l2.16 2.16C10.74 7.13 11.35 7 12 7zM2 4.27l2.28 2.28.46.46C3.08 8.3 1.78 10.02 1 12c1.73 4.39 6 7.5 11 7.5 1.55 0 3.03-.3 4.38-.84l.42.42L19.73 22 21 20.73 3.27 3 2 4.27zM7.53 9.8l1.55 1.55c-.05.21-.08.43-.08.65 0 1.66 1.34 3 3 3 .22 0 .44-.03.65-.08l1.55 1.55c-.67.33-1.41.53-2.2.53-2.76 0-5-2.24-5-5 0-.79.2-1.53.53-2.2zm4.31-.78l3.15 3.15.02-.16c0-1.66-1.34-3-3-3l-.17.01z\\"></path></svg></div><input type=\\"password\\" aria-invalid=\\"false\\" autoComplete=\\"current-password\\" id=\\"password\\" required=\\"\\" class=\\"MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart\\"/><fieldset aria-hidden=\\"true\\" class=\\"PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline\\"><legend class=\\"PrivateNotchedOutline-legendLabelled-3 PrivateNotchedOutline-legendNotched-4\\"><span>Password *</span></legend></fieldset></div></div></div></div></div></div><div class=\\"MuiGrid-root MuiGrid-item\\"><div class=\\"MuiCardActions-root MuiCardActions-spacing\\"><p class=\\"MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom\\"><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained width align MuiButton-containedPrimary MuiButton-fullWidth\\" tabindex=\\"0\\" type=\\"button\\" align=\\"center\\"><span class=\\"MuiButton-label\\">Sign In</span></button></p></div></div></div></div></div>"`;
